### PR TITLE
For Loops with Grouped NDRange

### PR DESCRIPTION
**Pull request by xumingkuan**
Related issue id = #615

Debugging now.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by yuanming-hu** [comment]
(I would appreciate it if you add a `[skip ci]` prefix to your commit message to avoid unnecessary CI builds. Thanks!)

**Comment by xumingkuan** [comment]
Can I place a new `ti.Vector` inside a kernel?
https://github.com/taichi-dev/taichi/blob/02a4048b547e0c00c65856f47351c4ecf0163268/python/taichi/lang/matrix.py#L63
Or can I use `ti.make_var_vector`? I don't know why this fails with `Windows fatal exception: access violation`.
```python
@ti.kernel
def test():
    if ti.static(1):
        I = ti.make_var_vector(size=2)
        for i, j in ti.ndrange((x0, y0), (x1, y1)):
            I[0] = i
            I[1] = j
            val[I] = I[0] + I[1] * 2
```

**Comment by yuanming-hu** [comment]
I think you can look at the struct-for implementation of `for I in ti.grouped(x)` in `transformer.py`

**Comment by archibate** [comment]
Try `ti.Vector([0, 0])`?

**Comment by xumingkuan** [comment]
It works. Thanks.

File: python/taichi/lang/transformer.py [hunk 1]
```diff
@@ -332,7 +333,9 @@ def visit_For(self, node):
         __I = __ndrange_I
             '''.format(node.target.id, dim)
             t = ast.parse(template).body[0]
-            t.body[0].value = node.iter if is_ndrange_for == 1 else node.iter.args[0]
+            t.body[
+                0].value = node.iter if is_ndrange_for == 1 else node.iter.args[
+                    0]
```
Comment (xumingkuan) [inline python/taichi/lang/transformer.py]: It looks bad here...

**Comment by yuanming-hu** [comment]
> Can I place a new `ti.Vector` inside a kernel?
> https://github.com/taichi-dev/taichi/blob/02a4048b547e0c00c65856f47351c4ecf0163268/python/taichi/lang/matrix.py#L63
> 
> 
> Or can I use `ti.make_var_vector`? I don't know why this fails with `Windows fatal exception: access violation`.
> ```python
> @ti.kernel
> def test():
>     if ti.static(1):
>         I = ti.make_var_vector(size=2)
>         for i, j in ti.ndrange((x0, y0), (x1, y1)):
>             I[0] = i
>             I[1] = j
>             val[I] = I[0] + I[1] * 2
> ```
> 
> (`I = [0, 0]` works here)

`I = ti.make_var_vector(size=2)` here you are coding before the frontend transform happens. This is different from emitting the same statement in the transform. Specifically, your code will be transformed into `I = ti.expr_init(ti.make_var_vector(size=2))`

**Comment by yuanming-hu** [comment]
@xumingkuan Sorry I have been too busy today day. Let me know if you want to briefly Skype chat. I'm available before 11:59 pm.

**Comment by xumingkuan** [comment]
The test about the new feature passes on my laptop. Slowly running other tests to figure out which fails now, and I think it unnecessary to chat today.

> `I = ti.make_var_vector(size=2)` here you are coding before the frontend transform happens. This is different from emitting the same statement in the transform. Specifically, your code will be transformed into `I = ti.expr_init(ti.make_var_vector(size=2))`

Yes, but
```
'''
if ti.static(1):
    __ndrange = 0
    {} = ti.make_var_vector(size={})
    for __ndrange_I in range(0):
        __I = __ndrange_I
            '''.format(node.target.id, dim)
```
somehow didn't work, and
```
'''
if ti.static(1):
    __ndrange = 0
    {} = ti.Vector([0] * {})
    for __ndrange_I in range(0):
        __I = __ndrange_I
            '''.format(node.target.id, dim)
```
worked for me.

> ```python
> @ti.kernel
> def test():
>     if ti.static(1):
>         I = ti.make_var_vector(size=2)
>         for i, j in ti.ndrange((x0, y0), (x1, y1)):
>             I[0] = i
>             I[1] = j
>             val[I] = I[0] + I[1] * 2
> ```

This was a failed attempt to generate similar preprocessed statements without `ti.grouped`.


Actually I don't get the point about
https://github.com/taichi-dev/taichi/blob/02a4048b547e0c00c65856f47351c4ecf0163268/python/taichi/lang/expr.py#L373
Does the empty string mean the variables don't have their names?

I think `ti.Vector([0, 0])` is more intuitive.

File: python/taichi/lang/transformer.py (hunk 1 above)
Comment (yuanming-hu) [reply python/taichi/lang/transformer.py]: Yeah, we should probably break it down into shorter lines :-)

**Comment by yuanming-hu** [comment]
> Does the empty string mean the variables don't have their names?

Yeah, they don't have a name. Taichi IR uses `IdExpression` instead of the strings to distinguish variables.

I think you can just stick to whatever works. Thanks.

**Comment by xumingkuan** [comment]
The reason CI failed was my attempt to change (at least partially) the code style in the string to 4-space indents. Should be fixed now.

**Comment by xumingkuan** [comment]
Will debug tomorrow.

**Comment by yuanming-hu** [comment]
Probably someone misclicked: https://github.com/taichi-dev/taichi/issues/621

File: python/taichi/lang/transformer.py [hunk 2]
```diff
@@ -278,64 +278,108 @@ def visit_For(self, node):
         if node.orelse:
             raise TaichiSyntaxError(
                 "'else' clause for 'for' not supported in Taichi kernels")
-        decorated = isinstance(node.iter, ast.Call) and isinstance(
-            node.iter.func, ast.Attribute) and isinstance(node.iter.func.value, ast.Name) \
-                    and node.iter.func.value.id == 'ti'
-        is_ndrange_for = False
-        is_static_for = False
-        is_grouped = False
+        decorated = isinstance(node.iter, ast.Call) \
```
Comment (yuanming-hu) [inline python/taichi/lang/transformer.py]: These lines are way too long - maybe it's worth extracting some common patterns into a (static) member method :-)

**Comment by xumingkuan** [comment]
Tests passed on my laptop so I tried to `rebase master`, but git complained that 12abb69 introduced conflicts that can't be auto-merged. So I cherry-picked new commits on master, but it resulted in a large change set...

**Comment by yuanming-hu** [comment]
Sorry I have to rush for a meeting soon but I believe you can work this out on your own :-)

You can always force-push to your own branch to get rid of the formatting commit.

**Comment by xumingkuan** [comment]
How to test python 3.8 on Windows?

**Comment by yuanming-hu** [comment]
Cool pls append `[release]` to your commit msg

**Comment by xumingkuan** [comment]
The previous build had only python 3.8 errored, but this build had only python 3.6 errored...?

**Comment by xumingkuan** [comment]
AppVeyor build succeeded just now. Maybe it's another cross-platform bug?

**Comment by yuanming-hu** [comment]
Yeah, it seems to me that Travis failed to do anything in that Python 3.6 build. Is this ready for review?

**Comment by xumingkuan** [comment]
Yeah it's ready for review.

File: tests/python/test_grouped.py [hunk 3]
```diff
@@ -31,7 +31,7 @@ def test():
 
 @ti.all_archs
 def test_grouped():
-    ti.get_runtime().print_preprocessed = True
```
Comment (yuanming-hu) [inline tests/python/test_grouped.py]: Please clean up commented debugging options in the test files.

File: tests/python/test_grouped.py [hunk 4]
```diff
@@ -53,3 +53,30 @@ def test():
...
+    y1 = 6
+
+    @ti.kernel
+    def test():
+        for I in ti.grouped(ti.ndrange((x0, y0), (x1, y1))):
+            val[I] = I[0] + I[1] * 2
+
+    test()
+
+    for i in range(x0, y0):
```
Comment (yuanming-hu) [inline tests/python/test_grouped.py]: Better test entries outside `ti.ndrange((x0, y0), (x1, y1))` being zero?

File: tests/python/test_ad_demote_dense.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 
 @ti.archs_excluding(ti.metal, ti.opengl)
 def test_ad_demote_dense():
+    ti.get_runtime().print_preprocessed = True
```
Comment (yuanming-hu) [inline tests/python/test_ad_demote_dense.py]: Please remove this line. Thanks.

File: python/taichi/lang/transformer.py [hunk 6]
```diff
@@ -278,64 +278,102 @@ def visit_For(self, node):
         if node.orelse:
             raise TaichiSyntaxError(
                 "'else' clause for 'for' not supported in Taichi kernels")
-        decorated = isinstance(node.iter, ast.Call) and isinstance(
-            node.iter.func, ast.Attribute) and isinstance(node.iter.func.value, ast.Name) \
-                    and node.iter.func.value.id == 'ti'
-        is_ndrange_for = False
-        is_static_for = False
-        is_grouped = False
+        def is_decorated(iter):
```
Comment (yuanming-hu) [inline python/taichi/lang/transformer.py]: Cool!

File: python/taichi/lang/transformer.py [hunk 7]
```diff
@@ -278,64 +278,102 @@ def visit_For(self, node):
...
             elif attr.attr == 'ndrange':
-                is_ndrange_for = True
+                is_ndrange_for = 1
             else:
                 raise Exception('Not supported')
+            if double_decorated:
+                attr = node.iter.args[0].func
+                # inner decorator
+                if attr.attr == 'static':
+                    is_static_for = 2
```
Comment (yuanming-hu) [inline python/taichi/lang/transformer.py]: One thing to improve here: if the user does `for i in ti.static(ti.static(range(x))`, we should throw a `TaichiSyntaxError`.

**Comment by yuanming-hu** [comment]
Great job @xumingkuan. Now everything looks really nice. I fixed a few corner cases (e.g.  `for I in ti.grouped(ti.ndrange())` and error on other nested constructs.)
I'll merge it in once CI passes.

Have a good weekend!

**Pull request merged**